src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.cpp

```diff
@@ -1,26 +1,90 @@
 #include "MetadataStorageFromPlainObjectStorage.h"
 #include <Disks/IDisk.h>
+#include <Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.h>
 #include <Disks/ObjectStorages/StaticDirectoryIterator.h>
+
+#include <IO/ReadHelpers.h>
+#include <Common/StringUtils/StringUtils.h>
 #include <Common/filesystemHelpers.h>
+#include <Common/getRandomASCIIString.h>
 #include <Common/logger_useful.h>
-#include <Common/StringUtils/StringUtils.h>
-#include <IO/WriteHelpers.h>
+#include "CommonPathPrefixKeyGenerator.h"
 
+#include <filesystem>
+#include <ranges>
+#include <tuple>
 
 namespace DB
 {
 
-MetadataStorageFromPlainObjectStorage::MetadataStorageFromPlainObjectStorage(
-    ObjectStoragePtr object_storage_,
-    String storage_path_prefix_)
+namespace
+{
+
+constexpr auto PREFIX_PATH_FILE_NAME = "prefix.path";
+
+std::filesystem::path normalizePath(const std::filesystem::path & path)
+{
+    return std::filesystem::path(path).lexically_normal();
+}
+
+std::filesystem::path normalizeDirectoryPath(const std::filesystem::path & path)
+{
+    return normalizePath(path) / "";
+}
+
+MetadataStorageFromPlainObjectStorage::PathMap loadPathPrefixMap(const std::string & root, ObjectStoragePtr object_storage)
+{
+    MetadataStorageFromPlainObjectStorage::PathMap result;
+
+    RelativePathsWithMetadata files;
+    object_storage->listObjects(root, files, 0);
```

Comment:
It will be better to use `IObjectStorage::iterate` instead of `IObjectStorage::listObjects`, because:
1. it will be more memory efficient, the number of files we list can be really huge
2. we could parallelize loading of metadata on server startup,  `IObjectStorageIterator` is thread-safe. Let's add a setting in disk configuration like `load_metadata_threads` and use threadpool to load metadata?

Reply:
Is it tested with minio?
i'm getting
```
2024.04.26 02:28:04.184442 [ 357825 ] {} <Error> Application: Code: 499. DB::Exception: Could not list objects in bucket 'test' with prefix 'clickhouse-disks/', S3 exception: `InvalidArgument`, message: 'Unable to parse ExceptionName: InvalidArgument Message: The continuation token provided is incorrect'. (S3_ERROR), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x000000001349ac72
1. ./build/./src/Common/Exception.cpp:101: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000afa7939
2. ./contrib/llvm-project/libcxx/include/string:1499: DB::S3Exception::S3Exception<String, String, String, String>(Aws::S3::S3Errors, fmt::v8::basic_format_string<char, fmt::v8::type_identity<String>::type, fmt::v8::type_identity<String>::type, fmt::v8::type_identity<String>::type, fmt::v8::type_identity<String>::type>, String&&, String&&, String&&, String&&) @ 0x000000000ea32922
3. ./build/./src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp:147: DB::(anonymous namespace)::S3IteratorAsync::getBatchAndCheckNext(std::vector<DB::RelativePathWithMetadata, std::allocator<DB::RelativePathWithMetadata>>&) @ 0x000000000f3d2d7f
4. ./build/./src/Disks/ObjectStorages/ObjectStorageIteratorAsync.cpp:70: DB::IObjectStorageIteratorAsync::BatchAndHasNext std::__function::__policy_invoker<DB::IObjectStorageIteratorAsync::BatchAndHasNext ()>::__call_impl<std::__function::__default_alloc_func<DB::IObjectStorageIteratorAsync::scheduleBatch()::$_0, DB::IObjectStorageIteratorAsync::BatchAndHasNext ()>>(std::__function::__policy_storage const*) @ 0x000000000eaf2ae6
5. ./contrib/llvm-project/libcxx/include/__functional/function.h:0: ? @ 0x000000000eaef6d4
6. ./contrib/llvm-project/libcxx/include/future:1380: std::packaged_task<DB::IObjectStorageIteratorAsync::BatchAndHasNext ()>::operator()() @ 0x000000000eaefc56
7. ./base/base/../base/wide_integer_impl.h:817: ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>) @ 0x000000000b0519cc
8. ./build/./src/Common/ThreadPool.cpp:0: ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() @ 0x000000000b0547f1
9. ./base/base/../base/wide_integer_impl.h:817: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000b04f002
10. ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000b0533ee
11. ? @ 0x00007dbff9494ac3
```

Reply:
Ah, yes, this iterator is not used anywhere now, but it will be used for s3 table engine after https://github.com/ClickHouse/ClickHouse/pull/59767 is merged. It did not work there initially as well, but I made several fixes and it started to work. I will extract the fixes into a separate PR to faster merge them. If you want, you can copy them into your PR.

Reply:
I will leave it as-is for now, but follow up after either https://github.com/ClickHouse/ClickHouse/pull/59767 or the PR with essential fixes is merged.